HNT-561-address high security alerts

jwalterss opened the pull request:
What changed? What is the business/product goal?

Add resolutions to package.json to force specific versions:

`{`  
  `"resolutions": {`  
    `"body-parser": "1.20.3",`  
    `"path-to-regexp": "0.1.12",`  
    `"cross-spawn": "7.0.5"`  
  `}`  
`}`



## Implementation Decisions

- Uses pnpm package manager's resolution feature
- Updates are patch-level changes, minimizing breaking change risk
- All updates stay within their respective major versions
- Addresses DoS and ReDoS vulnerabilities


## References

JIRA ticket:

https://mozilla-hub.atlassian.net/browse/HNT-561


Documentation:

Critical security updates needed for several dependencies to address high-severity vulnerabilities while avoiding breaking changes.

## Affected Packages

### 1. body-parser

- **Current Version**: 1.20.2
- **Required Version**: >=1.20.3
- **Issue**: Vulnerable to denial of service when URL encoding is enabled
- **Path**: Via express@4.19.2 in multiple dependencies

### 2. path-to-regexp

- **Current Version**: 0.1.7
- **Required Version**: >=0.1.12
- **Issues**:
    - Unpatched ReDoS in 0.1.x
    - Outputs backtracking regular expressions
- **Path**: Via express@4.19.2 in multiple dependencies

### 3. cross-spawn

- **Current Version**: 7.0.3
- **Required Version**: >=7.0.5
- **Issue**: Regular Expression Denial of Service (ReDoS)
- **Path**: Multiple paths including tsup@8.2.4 and cdktf-cli@0.20.8

## Required Actions

1. Add resolutions to package.json
2. Run `pnpm install`
3. Run test suite to verify no breaking changes
4. Test application functionality, especially:
    - Express.js routes and API endpoints
    - File system operations
    - Child process spawning

## Notes

- Updates are security-focused patches
- No major version changes involved
- Minimal risk of breaking changes due to patch-level updates
- Full test coverage recommended before deployment

## References

- body-parser: [https://github.com/advisories/GHSA-qwcr-r2fm-qrc7](https://github.com/advisories/GHSA-qwcr-r2fm-qrc7)
- path-to-regexp: [https://github.com/advisories/GHSA-rhx6-c78j-4q9w](https://github.com/advisories/GHSA-rhx6-c78j-4q9w)
- cross-spawn: [https://github.com/advisories/GHSA-3xgq-45jj-v275](https://github.com/advisories/GHSA-3xgq-45jj-v275)

jwalterss commented:
Used targeted resolution express>path-to-regexp to only update the vulnerable Express dependency while preserving MSW's newer version (6.3.0) which uses a different API. This ensures our tests continue to work correctly.

jwalterss commented on package.json:
```diff
@@ -33,5 +33,10 @@
     "tsconfig": "workspace:*",
     "turbo": "^2.1.1"
+  },
+  "resolutions": {
```
From Slack:

So our vulnerabilities come through this chain: @apollo/server@4.11.0 → express@4.19.2 → [body-parser@1.20.2, path-to-regexp@0.1.7] . The main roadblock is that MSW 2.4.7 uses path-to-regexp@6.3.0 and Express 4.x uses path-to-regexp@0.1.7 which are completely different in function signature and pattern syntax. So MSW continues using its required version (6.3.0) while Express gets the security patch (0.1.12)
The 4.x series (even 4.21.x) still uses [path-to-regexp@0.1.x](mailto:path-to-regexp@0.1.x). Express 5.x would fix this but it's still in beta and would be a breaking change for Apollo Server. I took the targeted approach and applied patch-level security fixes that don't change any APIs.
I think that when Apollo Server 5.x is released or they have an update to support Express 5.x we should make that update and remove these resolutions.

mmiermans replied:
Would you mind putting it under 'Implementation Decisions' in the PR description?

jwalterss commented on package.json:
```diff
@@ -33,5 +33,10 @@
     "turbo": "^2.1.1"
+  },
+  "resolutions": {
+    "body-parser": "1.20.3",
+    "express>path-to-regexp": "0.1.12",
+    "cross-spawn": "7.0.5"
```
There's no harm in doing so!

mmiermans replied:
Actually, I think you need `~` in this case, to only allow patches. To be honest, I never understood the difference between `^` and `~`. This article explains it: https://nodesource.com/blog/semver-tilde-and-caret.

jwalterss commented:
What Changed? What is the business/product goal?
Fixed cdktf version conflicts across infrastructure packages to enable successful builds and deployments.
Root Cause: Multiple infrastructure packages were using different versions of core dependencies (cdktf, constructs, @cdktf/provider-*), causing TypeScript type conflicts and build failures.
Implementation Decisions

Standardized all cdktf-related dependencies to consistent versions across all packages
Updated constructs library to version 10.4.2 (matching terraform-modules requirement)
Removed unnecessary package resolutions that were no longer needed
Used direct dependency updates rather than resolutions for cleaner dependency management

Package Versions Standardized

cdktf: 0.20.8 → 0.20.11
cdktf-cli: 0.20.8 → 0.20.11
constructs: 10.3.0 → 10.4.2
@cdktf/provider-aws: Various versions → 19.49.1
@cdktf/provider-local: 10.1.0 → 10.1.1
@cdktf/provider-null: 10.0.0 → 10.0.1
@cdktf/provider-pagerduty: 13.11.6 → 13.15.1
@cdktf/provider-archive: 10.2.0 → 10.3.0

Affected Packages

infrastructure/prospect-translation-lambda
infrastructure/corpus-scheduler-lambda
infrastructure/[other-infrastructure-package]
packages/infrastructure-common

Changes Made

Updated package.json dependencies in all infrastructure packages
Aligned versions with @pocket-tools/terraform-modules@5.20.0 requirements
Removed cross-spawn resolution from root package.json (no longer needed)
Ensured consistent constructs version across monorepo

Validation

Local pnpm run synth passes for all infrastructure packages
CI builds now succeed without type conflicts
No breaking changes to functionality

References

JIRA ticket: https://mozilla-hub.atlassian.net/browse/HNT-561
Issue: TypeScript type conflicts preventing infrastructure deployments

jwalterss commented:
after running pnpm audit: 

7 vulnerabilities found
Severity: 5 low | 2 moderate

mmiermans reviewed (approved):
Thanks for going the extra mile to update the packages!

We tested this in the dev branch, to ensure the `terraform apply` succeeds and the dev curation tools look normal.
